Update the documentation for links.

vpetrovykh opened the pull request:
Cover in more detials the one-to-one, one-to-many, and many-to-many
scenarios as well as the use of backward link traversal.

Fixes #967

elprans commented on docs/datamodel/links.rst:
```diff
@@ -14,16 +14,265 @@ There are two kinds of link item declarations: *abstract links*, and
+:ref:`constraints <ref_datamodel_constraints>`.  Concrete links
+are defined on specific object types.
 
 Links are directional and have a *source* and one or more *targets*.
```
Links don't have multiple targets.  Rather, a single *source* may have multiple *links* of a kind.  It is easy to see the difference when you take link properties into account.

vpetrovykh replied:
OK, I reworded it to talk about sources having links of certain kinds.

elprans commented on docs/datamodel/links.rst:
```diff
@@ -14,16 +14,265 @@ There are two kinds of link item declarations: *abstract links*, and
+One-to-One
+----------
+
+A *one-to-one* relationship represents a situation where everything
+is broken up into pairs and all the pairs are unique. For example,
+``Employee`` and ``ReservedParking``:
```
```suggestion
A *one-to-one* relationship represents a situation where one object from 
a source set is linked to only one object in the target set, and vice versa.
For example, ``Employee`` and ``ReservedParking``:
```

vpetrovykh replied:
Updated

elprans commented on docs/datamodel/links.rst:
```diff
@@ -14,16 +14,265 @@ There are two kinds of link item declarations: *abstract links*, and
+
+Backward link traversal requires to specify the original link's source
+type, but other than that it works the same way as forward traversal.
+
```
What about `Many-to-One`?

vpetrovykh replied:
Many-to-one is mentioned in a backward link traversal description of one-to-many.

elprans replied:
That isn't enough.  `Many-to-One` is actually the most used option, so it deserves a top level mention.  Perhaps even before `One-to-One`.

elprans commented on docs/datamodel/links.rst:
```diff
@@ -14,16 +14,265 @@ There are two kinds of link item declarations: *abstract links*, and
+        required property name -> str {
+            constraint exclusive;
+        }
+        multi link cars -> Shoes {
```
cars?

vpetrovykh replied:
An artifact from various iterations. Good catch :)

elprans commented on docs/datamodel/links.rst:
```diff
@@ -14,16 +14,265 @@ There are two kinds of link item declarations: *abstract links*, and
+        }
+    }
+
+A ``Person`` can have potentially have more than one pair of
```
I think `Shoes` is an inconvenient example in the context of the English language, because you use the term "pair of ``Shoes``" when talking about a single object, so this might be somewhat confusing to non-native speakers.

vpetrovykh replied:
Yeah. I'm looking for something that would be exclusively owned in a one-to-many way. So that's why I went for a personal item.

Maybe a ``Shirt`` or a ``Phone`` then? With ``Phone`` the example of getting multiple new ones may be a bit weird.

vpetrovykh commented:
I've updated the order of presentation to be: many-to-one, one-to-one, many-to-many.
I've also updated the schema to use `Shirt` for the many-to-one example.

elprans commented on docs/datamodel/links.rst:
```diff
@@ -14,16 +14,236 @@ There are two kinds of link item declarations: *abstract links*, and
+    }
+
+It's possible to include both links ``owner`` and ``shirts`` to a
+schema, making one of them a computable expression in terms of the
```
Perhaps add a link to some section that discusses computables, or, better, add an example?  In fact, putting backward traversal into a computable link is a common pattern.

vpetrovykh replied:
Fair point. I might as well use this opportunity to show a version of the schema with the computable link. I think we also have a section where we talk about computables, so I can link to that as well.

elprans reviewed (approved):
LGTM, but please expand a bit on computables.

vpetrovykh commented:
I updated the wording for computables.

tailhook commented on docs/datamodel/links.rst:
```diff
@@ -14,16 +14,259 @@ There are two kinds of link item declarations: *abstract links*, and
+            constraint exclusive;
+        }
+    }
+    type Shoes {
```
s/Shoes/Shirt/

vpetrovykh replied:
Replaced

tailhook commented on docs/datamodel/links.rst:
```diff
@@ -14,16 +14,259 @@ There are two kinds of link item declarations: *abstract links*, and
+    SELECT M.<likes[IS Person] {
+        name
+    };
+
```
Somewhere here is a good place to introduce link properties?

vpetrovykh replied:
Link properties are a messy topic all of their own. We have a few places scattered across the docs touching on them, but perhaps they should be reorganized.

tailhook commented on docs/datamodel/links.rst:
```diff
@@ -14,16 +14,259 @@ There are two kinds of link item declarations: *abstract links*, and
+            constraint exclusive;
+        }
+        multi link shirts -> Shirt {
+            constraint exclusive;
```
I would put some stress on `exclusive` here. This is what makes `many-to-one` instead of `many-to-many` in this case. Otherwise it's easy to miss when making your own schema.

vpetrovykh replied:
I've added a comment to the schema.
